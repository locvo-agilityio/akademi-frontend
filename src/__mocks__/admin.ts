export const MOCK_ADMIN = {
  id: 1,
  name: 'Nabila A',
  role: 'Admin',
  avatar: 'https://bit.ly/dan-abramov',
};
