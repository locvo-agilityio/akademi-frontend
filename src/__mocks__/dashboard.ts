import {
  CalendarIcon,
  FoodIcon,
  StudentIcon,
  TeacherIcon,
} from '@/components/icons';

export const MOCK_SUMMARY_DASHBOARD = [
  {
    title: 'Students',
    total: 932,
    bgIcon: 'primary',
    icon: StudentIcon,
  },
  {
    title: 'Teachers',
    total: 754,
    bgIcon: 'secondary',
    icon: TeacherIcon,
  },
  {
    title: 'Events',
    total: 40,
    bgIcon: 'yellow',
    icon: CalendarIcon,
  },
  {
    title: 'Foods',
    total: 32,
    bgIcon: 'darkBlue',
    icon: FoodIcon,
  },
];
