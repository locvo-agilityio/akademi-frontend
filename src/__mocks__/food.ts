export const MOCK_FOODS = [
  {
    name: 'Beef Steak with Fried Potato',
    content: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit.',
    image:
      'https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcRCQPUcFz7CxVScB-zxL5pl2w6f4ntjw4mV-w&s',
  },
  {
    name: 'Beef Steak with Fried Potato New',
    content: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit.',
    image:
      'https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcRCQPUcFz7CxVScB-zxL5pl2w6f4ntjw4mV-w&s',
  },
];
