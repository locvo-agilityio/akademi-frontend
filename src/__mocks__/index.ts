export * from './admin';
export * from './dashboard';
export * from './student';
export * from './pagination';
export * from './message';
export * from './food';
export * from './teacher';
export * from './schedule';
export * from './payment';
