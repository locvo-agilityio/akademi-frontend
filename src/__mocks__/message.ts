export const MOCK_MESSAGES = [
  {
    name: 'Samantha William',
    content: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit.',
    avatar: 'https://bit.ly/dan-abramov',
    time: '12:45 PM',
  },
  {
    name: 'Tony Soap',
    content: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit.',
    avatar: 'https://bit.ly/dan-abramov',
    time: '12:45 PM',
  },
  {
    name: 'Jordan Nico',
    content: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit.',
    avatar: 'https://bit.ly/dan-abramov',
    time: '12:45 PM',
  },
  {
    name: 'Nadila Adja',
    content: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit.',
    avatar: 'https://bit.ly/dan-abramov',
    time: '12:45 PM',
  },
];
