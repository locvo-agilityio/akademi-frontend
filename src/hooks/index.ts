export * from './useStudent';
export * from './usePagination';
export * from './useTeacher';
export * from './useUpload';
