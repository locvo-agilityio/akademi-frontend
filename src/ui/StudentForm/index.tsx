import {
  Button,
  Flex,
  FormControl,
  FormLabel,
  HStack,
  Radio,
  RadioGroup,
  Stack,
  Text,
  useToast,
  VStack,
} from '@chakra-ui/react';
import { Controller, useForm } from 'react-hook-form';
import { useQueryClient } from '@tanstack/react-query';
import { Link, useNavigate } from 'react-router-dom';
import { useActionState, useCallback, useEffect, useTransition } from 'react';

// Components
import {
  CustomToast,
  InputField,
  TextareaField,
  UploadImage,
} from '@/components';

// Constants
import {
  CLASS_NAME,
  CLASS_NAME_OPTIONS,
  ERROR_MESSAGES,
  PUBLIC_ROUTERS,
  studentsQueryKeys,
  SUCCESS_MESSAGES,
  TOAST_STATUS,
  VALIDATION_RULES,
} from '@/constants';

// Types
import { IStudent } from '@/types';

// Utils
import {
  ADDITIONAL_PARENT_STUDENT_INFO,
  BASIC_PARENT_STUDENT_INFO,
  BASIC_STUDENT_INFO,
} from '@/utils';

// Actions
import { addStudent, editStudent } from '@/actions';

interface IStudentFormProps {
  defaultValues?: IStudent;
}

const StudentForm = ({ defaultValues }: IStudentFormProps) => {
  const toast = useToast();
  const queryClient = useQueryClient();
  const navigate = useNavigate();
  const [isPending, startTransition] = useTransition();
  const isEdit = !!defaultValues?.documentId;

  const actionForm = isEdit ? editStudent : addStudent;

  const [state, formAction] = useActionState(actionForm, undefined);
  const {
    firstName = '',
    lastName = '',
    parentName = '',
    phone = '',
    email = '',
    address = '',
    city = '',
    grade = CLASS_NAME.VII_A,
    avatar = '',
    birthday = '',
    parent = { firstName: '', lastName: '', email: '', phone: '', address: '' },
    payment = 'unpaid',
    documentId = '',
  } = defaultValues || {};

  const initialValues: IStudent = {
    firstName: firstName,
    lastName: lastName,
    parentName: parentName,
    phone: phone,
    email: email,
    address: address,
    city: city,
    grade: grade,
    avatar: avatar,
    birthday: birthday,
    parent: parent,
    payment: payment,
    documentId: documentId,
  };

  const {
    control,
    setValue,
    getValues,
    formState: { isValid, isDirty },
  } = useForm<IStudent>({
    mode: 'onBlur',
    reValidateMode: 'onChange',
    defaultValues: initialValues,
  });

  const handleFileChange = useCallback(
    (file: string) => {
      setValue('avatar', file, { shouldDirty: true });
    },
    [setValue],
  );

  const actionSubmit = useCallback(
    async (formData: FormData) => {
      try {
        startTransition(() => {
          formData.append('avatar', getValues('avatar'));

          if (isEdit)
            formData.append('documentId', defaultValues?.documentId as string);

          formAction(formData);

          queryClient.invalidateQueries({
            queryKey: studentsQueryKeys.lists(),
          });

          toast({
            render: () => (
              <CustomToast
                status={TOAST_STATUS.SUCCESS}
                message={
                  isEdit
                    ? SUCCESS_MESSAGES.EDIT_STUDENT_SUCCESS
                    : SUCCESS_MESSAGES.ADD_STUDENT_SUCCESS
                }
              />
            ),
          });
        });
      } catch (err) {
        toast({
          render: () => (
            <CustomToast
              status={TOAST_STATUS.ERROR}
              message={
                isEdit
                  ? ERROR_MESSAGES.EDIT_STUDENT_FAILED
                  : ERROR_MESSAGES.ADD_STUDENT_FAILED
              }
            />
          ),
        });
      }
    },
    [defaultValues?.documentId, getValues, isEdit, queryClient, toast],
  );

  useEffect(() => {
    if (state) {
      queryClient.setQueryData(
        studentsQueryKeys.detail(state.data.documentId),
        state,
      );

      navigate(PUBLIC_ROUTERS.STUDENTS);
    }
  }, [navigate, queryClient, state]);

  return (
    <VStack as="form" alignItems="flex-end" w="full" action={actionSubmit}>
      <VStack mt={5} w="full" bgColor="white" borderRadius="2xl">
        <Text
          as="h3"
          w="full"
          px={10}
          h="60px"
          borderTopRadius="2xl"
          bgColor="primary"
          color="white"
          size="lg"
          fontWeight="bold"
          display="flex"
          alignItems="center"
        >
          Student Details
        </Text>

        <HStack w="full" alignItems="flex-start" p={8} gap={10}>
          <Controller
            key="avatar"
            control={control}
            name="avatar"
            render={() => (
              <UploadImage imageUrl={avatar} onFileChange={handleFileChange} />
            )}
          />

          <Flex w="full" gap={7} direction="column" justifyContent="flex-start">
            <Controller
              key="firstName"
              control={control}
              rules={VALIDATION_RULES.FIRST_NAME}
              name="firstName"
              render={({ field, fieldState: { error } }) => (
                <InputField
                  label="First Name *"
                  placeholder="Please enter your first name"
                  {...field}
                  isError={!!error}
                  errorMessages={error?.message}
                  aria-label="first name"
                />
              )}
            />

            <Flex
              direction={{ base: 'column', xl: 'row' }}
              gap={6}
              alignItems="flex-end"
            >
              <Controller
                key="birthday"
                control={control}
                rules={VALIDATION_RULES.BIRTH_DATE}
                name="birthday"
                render={({ field, fieldState: { error } }) => (
                  <InputField
                    label="Date & Place of Birth *"
                    type="date"
                    max="2009-12-31"
                    placeholder="Please enter your birthday"
                    {...field}
                    isError={!!error}
                    errorMessages={error?.message}
                    aria-label="birthday"
                  />
                )}
              />

              <Controller
                key="city"
                control={control}
                rules={VALIDATION_RULES.CITY}
                name="city"
                render={({ field, fieldState: { error } }) => (
                  <InputField
                    placeholder="Please enter your city"
                    {...field}
                    isError={!!error}
                    errorMessages={error?.message}
                    aria-label="city"
                  />
                )}
              />
            </Flex>

            <Controller
              key="email"
              control={control}
              rules={VALIDATION_RULES.EMAIL}
              name="email"
              render={({ field, fieldState: { error } }) => (
                <InputField
                  label="Email *"
                  type="email"
                  placeholder="Please enter your email"
                  {...field}
                  isError={!!error}
                  errorMessages={error?.message}
                  aria-label="email"
                />
              )}
            />

            <Controller
              key="address"
              control={control}
              rules={VALIDATION_RULES.ADDRESS}
              name="address"
              render={({ field, fieldState: { error } }) => (
                <TextareaField
                  label="Address"
                  placeholder="Please enter your address"
                  {...field}
                  isError={!!error}
                  errorMessages={error?.message}
                  aria-label="address"
                />
              )}
            />
          </Flex>

          <Flex w="full" gap={7} direction="column" justifyContent="flex-start">
            {BASIC_STUDENT_INFO.map(
              ({ name, label, placeholder, type, rules, ariaLabel }) => (
                <Controller
                  key={name}
                  control={control}
                  rules={rules}
                  name={name}
                  render={({ field, fieldState: { error } }) => (
                    <InputField
                      label={label}
                      placeholder={placeholder}
                      type={type}
                      {...field}
                      isError={!!error}
                      errorMessages={error?.message}
                      aria-label={ariaLabel}
                    />
                  )}
                />
              ),
            )}
            <Controller
              key="grade"
              control={control}
              name="grade"
              render={({ field: { onChange, value } }) => (
                <FormControl>
                  <FormLabel
                    fontSize="md"
                    fontWeight="semibold"
                    marginInlineEnd={0}
                    minW="max-content"
                    color="darkBlue"
                  >
                    Class *
                  </FormLabel>

                  <RadioGroup
                    onChange={onChange}
                    value={value}
                    size="sm"
                    display="flex"
                    gap={10}
                  >
                    {CLASS_NAME_OPTIONS.map((option) => (
                      <Radio
                        key={option.value}
                        name={option.name}
                        value={option.value}
                        colorScheme={option.colorScheme}
                      >
                        <Text
                          p={2}
                          fontSize="xs"
                          color="white"
                          alignItems="center"
                          bgColor={option.bgColor}
                          justifyContent="center"
                          borderRadius="full"
                        >
                          {option.text}
                        </Text>
                      </Radio>
                    ))}
                  </RadioGroup>
                </FormControl>
              )}
            />
          </Flex>
        </HStack>
      </VStack>

      <VStack mt={5} w="full" bgColor="white" borderRadius="2xl">
        <Text
          as="h3"
          w="full"
          px={10}
          h="60px"
          borderTopRadius="2xl"
          bgColor="primary"
          color="white"
          size="lg"
          fontWeight="bold"
          display="flex"
          alignItems="center"
        >
          Parent Details
        </Text>

        <HStack w="full" alignItems="flex-start" p={8} gap={10}>
          <Flex w="full" gap={7} direction="column" justifyContent="flex-start">
            {BASIC_PARENT_STUDENT_INFO.map(
              ({ name, label, placeholder, type, rules, ariaLabel }) => (
                <Controller
                  key={name}
                  control={control}
                  rules={rules}
                  name={name}
                  render={({ field, fieldState: { error } }) => (
                    <InputField
                      label={label}
                      placeholder={placeholder}
                      type={type}
                      {...field}
                      isError={!!error}
                      errorMessages={error?.message}
                      aria-label={ariaLabel}
                    />
                  )}
                />
              ),
            )}

            <Controller
              key="address"
              control={control}
              rules={VALIDATION_RULES.ADDRESS}
              name="parent.address"
              render={({ field, fieldState: { error } }) => (
                <TextareaField
                  label="Address"
                  placeholder="Please enter your address"
                  {...field}
                  isError={!!error}
                  errorMessages={error?.message}
                  aria-label="address"
                />
              )}
            />
          </Flex>

          <Flex w="full" gap={7} direction="column" justifyContent="flex-start">
            {ADDITIONAL_PARENT_STUDENT_INFO.map(
              ({ name, label, placeholder, type, rules, ariaLabel }) => (
                <Controller
                  key={name}
                  control={control}
                  rules={rules}
                  name={name}
                  render={({ field, fieldState: { error } }) => (
                    <InputField
                      label={label}
                      placeholder={placeholder}
                      type={type}
                      {...field}
                      isError={!!error}
                      errorMessages={error?.message}
                      aria-label={ariaLabel}
                    />
                  )}
                />
              ),
            )}

            <Controller
              key="payment"
              control={control}
              name="payment"
              render={({ field: { onChange, value } }) => (
                <FormControl>
                  <FormLabel
                    fontSize="md"
                    fontWeight="semibold"
                    marginInlineEnd={0}
                    minW="max-content"
                    color="darkBlue"
                  >
                    Payments *
                  </FormLabel>

                  <RadioGroup onChange={onChange} value={value} size="sm">
                    <Stack direction="row">
                      <Radio
                        name="payment"
                        value="cash"
                        colorScheme="purple"
                        defaultChecked
                      >
                        Cash
                      </Radio>
                      <Radio name="payment" value="debit" colorScheme="purple">
                        Debit
                      </Radio>
                      <Radio name="payment" value="unpaid" colorScheme="purple">
                        Unpaid
                      </Radio>
                    </Stack>
                  </RadioGroup>
                </FormControl>
              )}
            />
          </Flex>
        </HStack>
      </VStack>

      <HStack mt={4} gap={3}>
        <Button
          as={Link}
          to={PUBLIC_ROUTERS.STUDENTS}
          w="full"
          p={7}
          variant="secondary"
          borderRadius="full"
        >
          Cancel
        </Button>

        <Button
          w="full"
          p={7}
          variant="primary"
          borderRadius="full"
          type="submit"
          isLoading={isPending}
          isDisabled={isPending || !isValid || !isDirty}
        >
          Submit
        </Button>
      </HStack>
    </VStack>
  );
};

export default StudentForm;
