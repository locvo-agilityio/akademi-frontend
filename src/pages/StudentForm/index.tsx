import { lazy, Suspense } from 'react';
import { useParams } from 'react-router-dom';

// Components
import { Fallback } from '@/components';

// Hooks
import { useGetStudent } from '@/hooks';

const StudentFormUI = lazy(() => import('@/ui/StudentForm/'));

const StudentForm = () => {
  const { id } = useParams();
  const { student, isStudentLoading } = useGetStudent(id ?? '');

  return (
    <Suspense fallback={<Fallback />}>
      {!isStudentLoading && <StudentFormUI defaultValues={student?.data} />}
    </Suspense>
  );
};

export default StudentForm;
