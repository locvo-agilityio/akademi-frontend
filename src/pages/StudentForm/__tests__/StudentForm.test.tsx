import { render } from '@testing-library/react';

// Components
import StudentForm from '..';

// Mocks
import { MOCK_STUDENTS } from '@/__mocks__';

// Hooks
import { useGetStudent } from '@/hooks';

jest.mock('@/hooks', () => ({
  useGetStudent: jest.fn(() => ({
    student: MOCK_STUDENTS[0],
    isStudentLoading: false,
  })),
}));

describe('StudentForm', () => {
  it('should render successfully', () => {
    const { container } = render(<StudentForm />);

    expect(container).toMatchSnapshot();
  });

  it('should render fallbackUI when loading', () => {
    (useGetStudent as jest.Mock).mockReturnValue({
      student: null,
      isStudentLoading: true,
    });

    const { container } = render(<StudentForm />);

    expect(container).toBeInTheDocument();
  });
});
