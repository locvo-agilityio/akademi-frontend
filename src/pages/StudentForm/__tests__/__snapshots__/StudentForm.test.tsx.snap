// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StudentForm should render successfully 1`] = `
<div>
  <div
    class="css-tkevr6"
  >
    <div
      class="chakra-spinner css-1y7joxr"
      title="Fallback"
    >
      <span
        class="css-8b45rq"
      >
        Loading...
      </span>
    </div>
  </div>
</div>
`;
